### PR TITLE
added `iso_oscar_gap` for univar. polyn. rings

### PR DESCRIPTION
**Pull request by ThomasBreuer**
This is part of what issue #791 asks for.

### CODE REVIEW

**Comment by fingolfin** [comment]
Conflict needs to be resolved

File: src/GAP/iso_gap_oscar.jl [hunk 1]
```diff
@@ -55,13 +59,26 @@ function _iso_gap_oscar_field_rationals(F::GAP.GapObj)
    return MapFromFunc(x -> fmpq(x), x -> GAP.Obj(x), F, QQ)
 end
 
+function _iso_gap_oscar_ring_integers(F::GAP.GapObj)
+   return MapFromFunc(x -> fmpz(x), x -> GAP.Obj(x), F, ZZ)
```
Comment (fingolfin) [inline src/GAP/iso_gap_oscar.jl:63]: Actually, that's an unnecessary level of indirection (same for the QQ code a few lines above -- all my fault, of course)
```suggestion
   return MapFromFunc(fmpz, GAP.Obj, F, ZZ)
```

File: src/GAP/iso_oscar_gap.jl [hunk 2]
```diff
@@ -211,6 +209,50 @@ end
...
+@attributes GFPPolyRing # TODO: port this to Nemo
+#TODO: support
+# `NmodPolyRing` (from `Nemo.NmodRing(UInt64(6))`)
+# `FmpzModPolyRing` (from `Nemo.FmpzModRing(fmpz(2))`)
+# `FqDefaultPolyRing` (from `FqDefaultFiniteField(fmpz(2), 3, :x)`)
+
+function _iso_oscar_gap_polynomial_ring_functions(RO::PolyRing{T}, RG::GAP.GapObj, coeffs_iso::MapFromFunc) where T
+   fam = GAP.Globals.ElementsFamily(GAP.Globals.FamilyObj(codomain(coeffs_iso)))
+   ind = GAP.Globals.IndeterminateNumberOfUnivariateRationalFunction(
+           GAP.Globals.IndeterminatesOfPolynomialRing(RG)[1])
```
Comment (fingolfin) [inline src/GAP/iso_oscar_gap.jl]: Let's help Julia deduce the right types
```suggestion
   fam = GAP.Globals.ElementsFamily(GAP.Globals.FamilyObj(codomain(coeffs_iso)))::GapObj
   ind = GAP.Globals.IndeterminateNumberOfUnivariateRationalFunction(
           GAP.Globals.IndeterminatesOfPolynomialRing(RG)[1])::GapObj
```

File: src/GAP/iso_oscar_gap.jl [hunk 3]
```diff
@@ -211,6 +209,50 @@ end
...
+
+   f = function(x::PolyElem{T})
+      cfs = GAP.GapObj([coeffs_iso(x) for x in coefficients(x)])
+      return GAP.Globals.UnivariatePolynomialByCoefficients(fam, cfs, ind)
```
Comment (fingolfin) [inline src/GAP/iso_oscar_gap.jl]:
```suggestion
      return GAP.Globals.UnivariatePolynomialByCoefficients(fam, cfs, ind)::::GapObj
```

File: src/GAP/iso_oscar_gap.jl [hunk 4]
```diff
@@ -211,6 +209,50 @@ end
...
+   end
+
+   finv = function(x)
+      GAP.Globals.IsPolynomial(x) || error("$x is not a GAP polynomial")
+      cfs = Vector{Any}(GAP.Globals.CoefficientsOfUnivariatePolynomial(x))
```
Comment (fingolfin) [inline src/GAP/iso_oscar_gap.jl]:
```suggestion
      cfs = Vector{Any}(GAP.Globals.CoefficientsOfUnivariatePolynomial(x)::::GapObj)
```

**Review by fingolfin — APPROVED** [review]
Looks good to me. I would help Juli a bit with the type inference (it can't know that these functions return `::GapObj`. Of course could also think about treating all of these to the `@wrap` treatment, with suitable return type annotations, then the code gets more compact and (slightly) more efficient.

File: src/GAP/iso_gap_oscar.jl (hunk 1 above)
Comment (ThomasBreuer) [reply src/GAP/iso_gap_oscar.jl:63]: No.
`MapFromFunc` expects two `Function` arguments, but `fmpz` and `GAP.Obj` have type `DataType`.

Comment (fingolfin) [reply src/GAP/iso_gap_oscar.jl:63]: Ah, that sounds like a deficiency in `MapFromFunc` then.

Comment (ThomasBreuer) [reply src/GAP/iso_gap_oscar.jl:63]: @fingolfin see thofma/Hecke.jl/pull/592

**Pull request merged**